# Tackling the Health Crises in Africa
People dying from lack of medical resources
## Introduction
The lack of access to adequate medical resources and facilities have led to the significant number of deaths. Some of these deaths could have been avoided by timely access to a medical professional or close proximity of hospitals.

Health systems across Africa are underfunded and understaffed. Less than half of African citizens (52%) about 615 million people have access to the health care they need, the quality of health services across the continent is generally poor, and family planning needs of half the continent’s women and girls are unmet.

**Let’s focus on Africa in this dataset**

You are required to provide solutions to the health challenges, especially in Africa. Ensure that you allow all your creative self to shine through and remember, Africa looks up to you for a solution.

## Data Wrangling

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

%matplotlib inline

In [2]:
os.listdir('./data')

['1. annual-number-of-deaths-by-cause.csv',
 '2. number-of-deaths-by-age-group.csv',
 '3. Medical Doctors Per 10000 population.xlsx',
 '4. ISO 3166_country-and-continent-codes-list-csv.csv',
 '5. World Population.csv',
 '6. Current health expenditure (% of GDP).xlsx',
 'World Healthcaredata.xlsx',
 '~$World Healthcaredata.xlsx']

### Assessment

In [3]:
causesofDeath_df = pd.read_csv('./data/1. annual-number-of-deaths-by-cause.csv')

In [4]:
causesofDeath_df

Entity Code  Year Number of executions (Amnesty International)  \
0     Afghanistan  AFG  2007                                           15   
1     Afghanistan  AFG  2008                                           17   
2     Afghanistan  AFG  2009                                            0   
3     Afghanistan  AFG  2011                                            2   
4     Afghanistan  AFG  2012                                           14   
...           ...  ...   ...                                          ...   
7268     Zimbabwe  ZWE  2015                                          NaN   
7269     Zimbabwe  ZWE  2016                                          NaN   
7270     Zimbabwe  ZWE  2017                                          NaN   
7271     Zimbabwe  ZWE  2018                                          NaN   
7272     Zimbabwe  ZWE  2019                                          NaN   

      Meningitis  Alzheimer's disease and other dementias  \
0         2933.0                                   1402.0   
1         2731.0                                   1424.0   
2         2460.0                                   1449.0   
3         2327.0                                   1508.0   
4         2254.0                                   1544.0   
...          ...                                      ...   
7268      1439.0                                    754.0   
7269      1457.0                                    767.0   
7270      1460.0                                    781.0   
7271      1450.0                                    795.0   
7272      1450.0                                    812.0   

      Parkinson's disease  Nutritional deficiencies  Malaria  Drowning  ...  \
0                   450.0                    2488.0    393.0    2127.0  ...   
1                   455.0                    2277.0    255.0    1973.0  ...   
2                   460.0                    2040.0    239.0    1852.0  ...   
3                   473.0                    1846.0    390.0    1775.0  ...   
4                   482.0                    1705.0     94.0    1716.0  ...   
...                   ...                       ...      ...       ...  ...   
7268                215.0                    3019.0   2518.0     770.0  ...   
7269                219.0                    3056.0   2050.0     801.0  ...   
7270                223.0                    2990.0   2116.0     818.0  ...   
7271                227.0                    2918.0   2088.0     825.0  ...   
7272                232.0                    2884.0   2068.0     827.0  ...   

      Chronic kidney disease  Poisonings  Protein  Terrorism (deaths)  \
0                     4490.0       512.0   2439.0              1199.0   
1                     4534.0       495.0   2231.0              1092.0   
2                     4597.0       483.0   1998.0              1065.0   
3                     4785.0       483.0   1805.0              1525.0   
4                     4846.0       482.0   1667.0              3521.0   
...                      ...         ...      ...                 ...   
7268                  2108.0       381.0   2990.0                 NaN   
7269                  2160.0       393.0   3027.0                 NaN   
7270                  2196.0       398.0   2962.0                 0.0   
7271                  2240.0       400.0   2890.0                 NaN   
7272                  2292.0       405.0   2855.0                 NaN   

      Road injuries  Chronic respiratory diseases  \
0            7425.0                        7222.0   
1            7355.0                        7143.0   
2            7290.0                        7045.0   
3            7432.0                        6916.0   
4            7494.0                        6878.0   
...             ...                           ...   
7268         2373.0                        2751.0   
7269         2436.0                        2788.0   
7270         2473.0                        2818.0 

In [5]:
causesofDeath_df.columns

Index(['Entity', 'Code', 'Year',
       'Number of executions (Amnesty International)', 'Meningitis',
       'Alzheimer's disease and other dementias', 'Parkinson's disease',
       'Nutritional deficiencies', 'Malaria', 'Drowning',
       'Interpersonal violence', 'Maternal disorders', 'HIV/AIDS',
       'Drug use disorders', 'Tuberculosis', 'Cardiovascular diseases',
       'Lower respiratory infections', 'Neonatal disorders',
       'Alcohol use disorders', 'Self', 'Exposure to forces of nature',
       'Diarrheal diseases', 'Environmental heat and cold exposure',
       'Neoplasms', 'Conflict and terrorism', 'Diabetes mellitus',
       'Chronic kidney disease', 'Poisonings', 'Protein', 'Terrorism (deaths)',
       'Road injuries', 'Chronic respiratory diseases',
       'Cirrhosis and other chronic liver diseases', 'Digestive diseases',
       'Fire, heat, and hot substances', 'Acute hepatitis'],
      dtype='object')

In [6]:
ageGroupDeaths_df = pd.read_csv("./data/2. number-of-deaths-by-age-group.csv")

In [7]:
ageGroupDeaths_df

Entity Code  Year  Deaths 70+ years  Deaths 50-69 years  \
0     Afghanistan  AFG  1990             34681               40624   
1     Afghanistan  AFG  1991             35271               41414   
2     Afghanistan  AFG  1992             35694               41829   
3     Afghanistan  AFG  1993             36433               42598   
4     Afghanistan  AFG  1994             37460               43923   
...           ...  ...   ...               ...                 ...   
6835     Zimbabwe  ZWE  2015             25573               32963   
6836     Zimbabwe  ZWE  2016             25804               33138   
6837     Zimbabwe  ZWE  2017             26030               33282   
6838     Zimbabwe  ZWE  2018             26279               33316   
6839     Zimbabwe  ZWE  2019             26666               33816   

      Deaths Age: 15-49 years  Deaths 5-14 years  Deaths Under 5  
0                       22968               5642           78170  
1                       26523               6433           83035  
2                       30000               6854           93859  
3                       32339               7206          106537  
4                       36817               8106          115334  
...                       ...                ...             ...  
6835                    43441               4506           26551  
6836                    42011               4592           25499  
6837                    40417               4737           24766  
6838                    38536               4430           23905  
6839                    38193               4282           23566  

[6840 rows x 8 columns]

In [8]:
doctorsPerPopulation_df = pd.read_excel('./data/3. Medical Doctors Per 10000 population.xlsx',
                                        header=2)

In [9]:
doctorsPerPopulation_df

IndicatorCode                                     Indicator  \
0          HWF_0001                  Medical doctors (per 10,000)   
1          HWF_0001                  Medical doctors (per 10,000)   
2          HWF_0001                  Medical doctors (per 10,000)   
3          HWF_0001                  Medical doctors (per 10,000)   
4          HWF_0001                  Medical doctors (per 10,000)   
...             ...                                           ...   
11503      HWF_0005  Medical doctors not further defined (number)   
11504      HWF_0005  Medical doctors not further defined (number)   
11505      HWF_0005  Medical doctors not further defined (number)   
11506      HWF_0005  Medical doctors not further defined (number)   
11507      HWF_0005  Medical doctors not further defined (number)   

      ParentLocationCode   ParentLocation ThreeLocCode Location  Period  \
0                    AFR           Africa          AGO   Angola    2018   
1                    AFR           Africa          AGO   Angola    2017   
2                    AFR           Africa          AGO   Angola    2009   
3                    AFR           Africa          AGO   Angola    2004   
4                    AFR           Africa          AGO   Angola    1997   
...                  ...              ...          ...      ...     ...   
11503                WPR  Western Pacific          WSM    Samoa    2010   
11504                WPR  Western Pacific          WSM    Samoa    2008   
11505                WPR  Western Pacific          WSM    Samoa    2005   
11506                WPR  Western Pacific          WSM    Samoa    2003   
11507                WPR  Western Pacific          WSM    Samoa    1999   

       FactValueNumeric Value  
0                  2.14  2.14  
1                  2.15  2.15  
2                  1.31  1.31  
3                  0.62  0.62  
4                  0.59  0.59  
...                 ...   ...  
11503             64.00    64  
11504             85.00    85  
11505             50.00    50  
11506             50.00    50  
11507            120.00   120  

[11508 rows x 9 columns]

In [10]:
doctorsPerPopulation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11508 entries, 0 to 11507
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   IndicatorCode       11508 non-null  object 
 1   Indicator           11508 non-null  object 
 2   ParentLocationCode  11508 non-null  object 
 3   ParentLocation      11508 non-null  object 
 4   ThreeLocCode        11508 non-null  object 
 5   Location            11508 non-null  object 
 6   Period              11508 non-null  int64  
 7   FactValueNumeric    11508 non-null  float64
 8   Value               11508 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 809.3+ KB


In [11]:
# let's find out the values that made 'Value' an object type
objectType = []

for i in doctorsPerPopulation_df.Value:
      try:
            float(i)
      except:
            objectType.append(i)

In [12]:
len(objectType)

4488

In [13]:
# converting to Series so that we can take a sample
objectType = pd.Series(objectType)
set(objectType.sample(10))

{'10\xa0498',
 '13\xa0272',
 '14\xa0464',
 '18\xa0697',
 '20\xa0670',
 '29\xa0537',
 '46\xa0726',
 '88\xa0070',
 '96\xa0038',
 '99\xa0455'}

Looks like some values have a whitespace character (`\xa0`)

In [14]:
doctorsPerPopulation_df.Value.sample(7)

5277     20 200
8018    112 699
4992     15 354
8783        494
6985       3306
4166     33 785
2218      27.63
Name: Value, dtype: object

In [15]:
doctorsPerPopulation_df.isna().sum()

IndicatorCode         0
Indicator             0
ParentLocationCode    0
ParentLocation        0
ThreeLocCode          0
Location              0
Period                0
FactValueNumeric      0
Value                 0
dtype: int64

In [16]:
countryCodes_df = pd.read_csv('./data/4. ISO 3166_country-and-continent-codes-list-csv.csv')

In [17]:
countryCodes_df

Continent_Name Continent_Code  \
0             Asia             AS   
1           Europe             EU   
2       Antarctica             AN   
3           Africa             AF   
4          Oceania             OC   
..             ...            ...   
249         Africa             AF   
250        Oceania             OC   
251           Asia             AS   
252           Asia             AS   
253           Asia             AS   

                                     Country_Name Two_Letter_Country_Code  \
0                Afghanistan, Islamic Republic of                      AF   
1                            Albania, Republic of                      AL   
2    Antarctica (the territory South of 60 deg S)                      AQ   
3        Algeria, People's Democratic Republic of                      DZ   
4                                  American Samoa                      AS   
..                                            ...                     ...   
249                           Zambia, Republic of                      ZM   
250                            Disputed Territory                      XX   
251                Iraq-Saudi Arabia Neutral Zone                      XE   
252                   United Nations Neutral Zone                      XD   
253                               Spratly Islands                      XS   

    Three_Letter_Country_Code  Country_Number  
0                         AFG             4.0  
1                         ALB             8.0  
2                         ATA            10.0  
3                         DZA            12.0  
4                         ASM            16.0  
..                        ...             ...  
249                       ZMB           894.0  
250                        XX             NaN  
251                        XE             NaN  
252                        XD             NaN  
253                        XS             NaN  

[254 rows x 6 columns]

In [18]:
population = pd.read_csv('./data/5. World Population.csv')

In [19]:
population

Entity Code   Year  Population (historical estimates)
0      Afghanistan  AFG -10000                              14737
1      Afghanistan  AFG  -9000                              20405
2      Afghanistan  AFG  -8000                              28253
3      Afghanistan  AFG  -7000                              39120
4      Afghanistan  AFG  -6000                              54166
...            ...  ...    ...                                ...
56908     Zimbabwe  ZWE   2017                           14751101
56909     Zimbabwe  ZWE   2018                           15052191
56910     Zimbabwe  ZWE   2019                           15354606
56911     Zimbabwe  ZWE   2020                           15669663
56912     Zimbabwe  ZWE   2021                           15993525

[56913 rows x 4 columns]

In [20]:
population.isna().sum()

Entity                                  0
Code                                 2662
Year                                    0
Population (historical estimates)       0
dtype: int64

In [21]:
population[population['Code'].isna()]

Entity Code   Year  \
259                           Africa  NaN -10000   
260                           Africa  NaN  -9000   
261                           Africa  NaN  -8000   
262                           Africa  NaN  -7000   
263                           Africa  NaN  -6000   
...                              ...  ...    ...   
54176  Upper-middle-income countries  NaN   2017   
54177  Upper-middle-income countries  NaN   2018   
54178  Upper-middle-income countries  NaN   2019   
54179  Upper-middle-income countries  NaN   2020   
54180  Upper-middle-income countries  NaN   2021   

       Population (historical estimates)  
259                               227636  
260                               323059  
261                               462991  
262                               670044  
263                               979129  
...                                  ...  
54176                         2500616723  
54177                         2516785800  
54178                         2530546049  
54179                         2541463877  
54180                         2548718303  

[2662 rows x 4 columns]

In [22]:
population[population['Code'].isna()].Entity.value_counts()

Africa                           259
Asia                             259
Europe                           259
High-income countries            259
Low-income countries             259
Lower-middle-income countries    259
North America                    259
Oceania                          259
South America                    259
Upper-middle-income countries    259
Saint Barthlemy                   72
Name: Entity, dtype: int64

In [23]:
healthExpenditure = pd.read_excel('./data/6. Current health expenditure (% of GDP).xlsx',
                                  header=4)

In [24]:
healthExpenditure

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                            Indicator Name      2000      2001      2002  \
0    Current health expenditure (% of GDP)       NaN       NaN       NaN   
1    Current health expenditure (% of GDP)  6.252428  6.390627  5.969416   
2    Current health expenditure (% of GDP)       NaN       NaN  9.443390   
3    Current health expenditure (% of GDP)  3.771302  3.770256  3.387634   
4    Current health expenditure (% of GDP)  1.908599  4.483516  3.329461   
..                                     ...       ...       ...       ...   
261  Current health expenditure (% of GDP)       NaN       NaN       NaN   
262  Current health expenditure (% of GDP)  4.666065  4.945151  4.832372   
263  Current health expenditure (% of GDP)  8.164721  7.939017  7.464463   
264  Current health expenditure (% of GDP)  7.152371  6.552485  6.794035   
265  Current health expenditure (% of GDP)       NaN       NaN       NaN   

         2003      2004      2005       2006  ...      2012      2013  \
0         NaN       NaN       NaN        NaN  ...       NaN       NaN   
1    6.635666  6.710734  6.528098   6.492506  ...  6.562794  6.565689   
2    8.941258  9.808474  9.948290  10.622766  ...  7.897176  8.805941   
3    4.681099  4.481882  4.277262   4.139578  ...  3.505065  3.621925   
4    3.547973  3.967201  2.852197   2.685537  ...  2.395750  2.732828   
..        ...       ...       ...        ...  ...       ...       ...   
261       NaN       NaN       NaN        NaN  ...       NaN       NaN   
262  5.757578  5.514405  5.205854   5.529299  ...  5.195936  4.960832   
263  7.915304  7.764534  7.689974   7.692216  ...  8.773108  8.855524   
264  7.189994  7.049950  6.863322   5.873673  ...  3.930529  4.690910   
265       NaN       NaN       NaN        NaN  ...  6.918354  7.110148   

         2014       2015       2016       2017       2018       2019  2020  \
0         NaN        NaN        NaN        NaN        NaN        NaN   NaN   
1    6.152063   6.482263   6.390654   6.317654   6.286829   6.266386   NaN   
2    9.528871  10.105348  11.818562  12.620817  14.126743  13.242202   NaN   
3    3.614749   3.834676   3.811248   3.789624   3.393160   3.366100   NaN   
4    2.434129   2.605795   2.713149   2.791503   2.540102   2.533360   NaN   
..        ...        ...        ...        ...        ...        ...   ...   
261       NaN        NaN        NaN        NaN        NaN        NaN   NaN   
262  4.841805   4.252858        NaN        NaN        NaN        NaN   NaN   
263  8.550915   8.790190   8.821429   8.722624   8.858297   9.109355   NaN   
264  3.829242   4.435102   4.477207   4.376977   5.032715   5.312203   NaN   
265  8.133525   7.452066   7.675163   7.469752   8.680062   7.703565   NaN   

     2021  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN  
..    ...  
261   NaN  
262   NaN  
263   NaN  
264   NaN  
265   NaN  

[266 rows x 25 columns]

In [25]:
healthExpenditure.isna().sum()

Country Name        0
Country Code        0
Indicator Name      0
2000               34
2001               34
2002               33
2003               31
2004               31
2005               31
2006               31
2007               31
2008               31
2009               31
2010               30
2011               29
2012               30
2013               31
2014               31
2015               31
2016               32
2017               31
2018               31
2019               32
2020              266
2021              266
dtype: int64

In [26]:
healthExpenditure['Indicator Name'].value_counts()

Current health expenditure (% of GDP)    266
Name: Indicator Name, dtype: int64

2020 and 2021 are both empty.

**Issues**:
      
`countryCodes_df`
* Split by `,` for the `Country_Name` column

`causesofDeath_df`
* ~~Create a subset of only African countries~~
* Create a column for 'Causes' and another for 'Case Count' using `.melt()`
* Drop unrelated death causes

`ageGroupDeaths_df`
* ~~Create a subset of only African countries~~
* Create a column for 'Age Range' and another for 'Death Count' using `.melt()`

`doctorsPerPopulation_df`
* ~~Create subset of only African countries~~
* Remove white space character (`\xa0`) from `Value` column

`healthExpenditure`
* ~~Create subset of African countries~~
* Drop 'Indicator Name'
* Create a column for 'Year' and 'Health Expenditure (% of GDP)' for values using `.melt()`

### Cleaning

In [27]:
causesofDeath_df_copy = causesofDeath_df.copy()
ageGroupDeaths_df_copy = ageGroupDeaths_df.copy()
doctorsPerPopulation_df_copy = doctorsPerPopulation_df.copy()
countryCodes_df_copy = countryCodes_df.copy()
healthExpenditure_copy = healthExpenditure.copy()

#### `countryCodes_df`
* Split by `,` for the `Country_Name` column

In [28]:
countryCodes_df_copy

Continent_Name Continent_Code  \
0             Asia             AS   
1           Europe             EU   
2       Antarctica             AN   
3           Africa             AF   
4          Oceania             OC   
..             ...            ...   
249         Africa             AF   
250        Oceania             OC   
251           Asia             AS   
252           Asia             AS   
253           Asia             AS   

                                     Country_Name Two_Letter_Country_Code  \
0                Afghanistan, Islamic Republic of                      AF   
1                            Albania, Republic of                      AL   
2    Antarctica (the territory South of 60 deg S)                      AQ   
3        Algeria, People's Democratic Republic of                      DZ   
4                                  American Samoa                      AS   
..                                            ...                     ...   
249                           Zambia, Republic of                      ZM   
250                            Disputed Territory                      XX   
251                Iraq-Saudi Arabia Neutral Zone                      XE   
252                   United Nations Neutral Zone                      XD   
253                               Spratly Islands                      XS   

    Three_Letter_Country_Code  Country_Number  
0                         AFG             4.0  
1                         ALB             8.0  
2                         ATA            10.0  
3                         DZA            12.0  
4                         ASM            16.0  
..                        ...             ...  
249                       ZMB           894.0  
250                        XX             NaN  
251                        XE             NaN  
252                        XD             NaN  
253                        XS             NaN  

[254 rows x 6 columns]

In [29]:
countryCodes_df_copy['Country_Name'] = countryCodes_df_copy['Country_Name'].apply(lambda x: x.split(',')[0])

In [30]:
countryCodes_df_copy.Country_Name.sample(15)

125     Liechtenstein
201      Sierra Leone
15            Armenia
147           Namibia
49            Mayotte
14         Bangladesh
117            Kuwait
4      American Samoa
94             Guyana
5             Andorra
198           Senegal
113            Jordan
138            Mexico
239          Tanzania
80            Georgia
Name: Country_Name, dtype: object

Let's get the subset for African countries

In [31]:
countryCodes_df_copy.Continent_Name.unique()

array(['Asia', 'Europe', 'Antarctica', 'Africa', 'Oceania',
       'North America', 'South America'], dtype=object)

In [32]:
africanCountries = countryCodes_df_copy.query('Continent_Name == "Africa"')
africanCountries = africanCountries.reset_index(drop=True)

In [33]:
africanCountries

Continent_Name Continent_Code              Country_Name  \
0          Africa             AF                   Algeria   
1          Africa             AF                    Angola   
2          Africa             AF                  Botswana   
3          Africa             AF                   Burundi   
4          Africa             AF                  Cameroon   
5          Africa             AF                Cape Verde   
6          Africa             AF  Central African Republic   
7          Africa             AF                      Chad   
8          Africa             AF                   Comoros   
9          Africa             AF                   Mayotte   
10         Africa             AF                     Congo   
11         Africa             AF                     Congo   
12         Africa             AF                     Benin   
13         Africa             AF         Equatorial Guinea   
14         Africa             AF                  Ethiopia   
15         Africa             AF                   Eritrea   
16         Africa             AF                  Djibouti   
17         Africa             AF                     Gabon   
18         Africa             AF                    Gambia   
19         Africa             AF                     Ghana   
20         Africa             AF                    Guinea   
21         Africa             AF             Cote d'Ivoire   
22         Africa             AF                     Kenya   
23         Africa             AF                   Lesotho   
24         Africa             AF                   Liberia   
25         Africa             AF    Libyan Arab Jamahiriya   
26         Africa             AF                Madagascar   
27         Africa             AF                    Malawi   
28         Africa             AF                      Mali   
29         Africa             AF                Mauritania   
30         Africa             AF                 Mauritius   
31         Africa             AF                   Morocco   
32         Africa             AF                Mozambique   
33         Africa             AF                   Namibia   
34         Africa             AF                     Niger   
35         Africa             AF                   Nigeria   
36         Africa             AF             Guinea-Bissau   
37         Africa             AF                   Reunion   
38         Africa             AF                    Rwanda   
39         Africa             AF              Saint Helena   
40         Africa             AF     Sao Tome and Principe   
41         Africa             AF                   Senegal   
42         Africa             AF                Seychelles   
43         Africa             AF              Sierra Leone   
44         Africa             AF                   Somalia   
45         Africa             AF              South Africa   
46         Africa             AF                  Zimbabwe   
47         Africa             AF               South Sudan   
48         Africa             AF            Western Sahara   
49         Africa             AF                     Sudan   
50         Africa             AF                 Swaziland   
51         Africa             AF                      Togo   
52         Africa             AF                   Tunisia   
53         Africa             AF                    Uganda   
54         Africa             AF                     Egypt   
55         Africa             AF                  Tanzania   
56         Africa             AF              Burkina Faso   
57         Africa             AF                    Zambia   

   Two_Letter_Country_Code Three_Letter_Country_Code  Country_Number  
0                       DZ                       DZA            12.0  
1                       AO                       AGO            24.0  
2                       BW                       BWA            72.0  
3                       BI                       BDI        

Convert Namibia's country code from NaN to 'NA'

In [34]:
africanCountries[africanCountries['Two_Letter_Country_Code'].isna()]

Continent_Name Continent_Code Country_Name Two_Letter_Country_Code  \
33         Africa             AF      Namibia                     NaN   

   Three_Letter_Country_Code  Country_Number  
33                       NAM           516.0

In [35]:
countryCodes_df_copy[countryCodes_df_copy['Two_Letter_Country_Code'].isna()]

Continent_Name Continent_Code Country_Name Two_Letter_Country_Code  \
147         Africa             AF      Namibia                     NaN   

    Three_Letter_Country_Code  Country_Number  
147                       NAM           516.0

In [36]:
africanCountries.iloc[33,3] = 'NA'
countryCodes_df_copy.iloc[147,3] = 'NA'

Looks like there are two Congos. The Democratic Republic of Congo (COD) and the Republic of Congo (COG). Let's convert change the name for COD only.

In [37]:
congoIndex = africanCountries.query('Three_Letter_Country_Code == "COD"').index
worldcongoIndex = countryCodes_df_copy.query('Three_Letter_Country_Code == "COD"').index

africanCountries.iloc[congoIndex,2] = 'Democratic Republic of Congo'
countryCodes_df_copy.iloc[worldcongoIndex,2] = 'Democratic Republic of Congo'

#### `causesofDeath_df`
* ~~Create a subset of only African countries~~
* Create a column for 'Causes' and another for 'Case Count' using `.melt()`
* Drop unrelated death causes

In [38]:
causesofDeath_df_copy

Entity Code  Year Number of executions (Amnesty International)  \
0     Afghanistan  AFG  2007                                           15   
1     Afghanistan  AFG  2008                                           17   
2     Afghanistan  AFG  2009                                            0   
3     Afghanistan  AFG  2011                                            2   
4     Afghanistan  AFG  2012                                           14   
...           ...  ...   ...                                          ...   
7268     Zimbabwe  ZWE  2015                                          NaN   
7269     Zimbabwe  ZWE  2016                                          NaN   
7270     Zimbabwe  ZWE  2017                                          NaN   
7271     Zimbabwe  ZWE  2018                                          NaN   
7272     Zimbabwe  ZWE  2019                                          NaN   

      Meningitis  Alzheimer's disease and other dementias  \
0         2933.0                                   1402.0   
1         2731.0                                   1424.0   
2         2460.0                                   1449.0   
3         2327.0                                   1508.0   
4         2254.0                                   1544.0   
...          ...                                      ...   
7268      1439.0                                    754.0   
7269      1457.0                                    767.0   
7270      1460.0                                    781.0   
7271      1450.0                                    795.0   
7272      1450.0                                    812.0   

      Parkinson's disease  Nutritional deficiencies  Malaria  Drowning  ...  \
0                   450.0                    2488.0    393.0    2127.0  ...   
1                   455.0                    2277.0    255.0    1973.0  ...   
2                   460.0                    2040.0    239.0    1852.0  ...   
3                   473.0                    1846.0    390.0    1775.0  ...   
4                   482.0                    1705.0     94.0    1716.0  ...   
...                   ...                       ...      ...       ...  ...   
7268                215.0                    3019.0   2518.0     770.0  ...   
7269                219.0                    3056.0   2050.0     801.0  ...   
7270                223.0                    2990.0   2116.0     818.0  ...   
7271                227.0                    2918.0   2088.0     825.0  ...   
7272                232.0                    2884.0   2068.0     827.0  ...   

      Chronic kidney disease  Poisonings  Protein  Terrorism (deaths)  \
0                     4490.0       512.0   2439.0              1199.0   
1                     4534.0       495.0   2231.0              1092.0   
2                     4597.0       483.0   1998.0              1065.0   
3                     4785.0       483.0   1805.0              1525.0   
4                     4846.0       482.0   1667.0              3521.0   
...                      ...         ...      ...                 ...   
7268                  2108.0       381.0   2990.0                 NaN   
7269                  2160.0       393.0   3027.0                 NaN   
7270                  2196.0       398.0   2962.0                 0.0   
7271                  2240.0       400.0   2890.0                 NaN   
7272                  2292.0       405.0   2855.0                 NaN   

      Road injuries  Chronic respiratory diseases  \
0            7425.0                        7222.0   
1            7355.0                        7143.0   
2            7290.0                        7045.0   
3            7432.0                        6916.0   
4            7494.0                        6878.0   
...             ...                           ...   
7268         2373.0                        2751.0   
7269         2436.0                        2788.0   
7270         2473.0                        2818.0 

In [39]:
# creating a subset of african countries
#african_three_letter_code = africanCountries.Three_Letter_Country_Code

#causesofDeath_df_copy = causesofDeath_df_copy.query('Code in @african_three_letter_code')

In [40]:
causesofDeath_df_copy.head()

Entity Code  Year Number of executions (Amnesty International)  \
0  Afghanistan  AFG  2007                                           15   
1  Afghanistan  AFG  2008                                           17   
2  Afghanistan  AFG  2009                                            0   
3  Afghanistan  AFG  2011                                            2   
4  Afghanistan  AFG  2012                                           14   

   Meningitis  Alzheimer's disease and other dementias  Parkinson's disease  \
0      2933.0                                   1402.0                450.0   
1      2731.0                                   1424.0                455.0   
2      2460.0                                   1449.0                460.0   
3      2327.0                                   1508.0                473.0   
4      2254.0                                   1544.0                482.0   

   Nutritional deficiencies  Malaria  Drowning  ...  Chronic kidney disease  \
0                    2488.0    393.0    2127.0  ...                  4490.0   
1                    2277.0    255.0    1973.0  ...                  4534.0   
2                    2040.0    239.0    1852.0  ...                  4597.0   
3                    1846.0    390.0    1775.0  ...                  4785.0   
4                    1705.0     94.0    1716.0  ...                  4846.0   

   Poisonings  Protein  Terrorism (deaths)  Road injuries  \
0       512.0   2439.0              1199.0         7425.0   
1       495.0   2231.0              1092.0         7355.0   
2       483.0   1998.0              1065.0         7290.0   
3       483.0   1805.0              1525.0         7432.0   
4       482.0   1667.0              3521.0         7494.0   

   Chronic respiratory diseases  Cirrhosis and other chronic liver diseases  \
0                        7222.0                                      3346.0   
1                        7143.0                                      3316.0   
2                        7045.0                                      3291.0   
3                        6916.0                                      3318.0   
4                        6878.0                                      3353.0   

   Digestive diseases  Fire, heat, and hot substances  Acute hepatitis  
0              6458.0                           481.0           3437.0  
1              6408.0                           462.0           3005.0  
2              6358.0                           448.0           2663.0  
3              6370.0                           448.0           2365.0  
4              6398.0                           445.0           2264.0  

[5 rows x 36 columns]

Replace NaN with 0s for 'Number of executions...'

In [41]:
causesofDeath_df_copy = causesofDeath_df_copy.fillna(0)

In [42]:
causesofDeath_df_copy.head()

Entity Code  Year Number of executions (Amnesty International)  \
0  Afghanistan  AFG  2007                                           15   
1  Afghanistan  AFG  2008                                           17   
2  Afghanistan  AFG  2009                                            0   
3  Afghanistan  AFG  2011                                            2   
4  Afghanistan  AFG  2012                                           14   

   Meningitis  Alzheimer's disease and other dementias  Parkinson's disease  \
0      2933.0                                   1402.0                450.0   
1      2731.0                                   1424.0                455.0   
2      2460.0                                   1449.0                460.0   
3      2327.0                                   1508.0                473.0   
4      2254.0                                   1544.0                482.0   

   Nutritional deficiencies  Malaria  Drowning  ...  Chronic kidney disease  \
0                    2488.0    393.0    2127.0  ...                  4490.0   
1                    2277.0    255.0    1973.0  ...                  4534.0   
2                    2040.0    239.0    1852.0  ...                  4597.0   
3                    1846.0    390.0    1775.0  ...                  4785.0   
4                    1705.0     94.0    1716.0  ...                  4846.0   

   Poisonings  Protein  Terrorism (deaths)  Road injuries  \
0       512.0   2439.0              1199.0         7425.0   
1       495.0   2231.0              1092.0         7355.0   
2       483.0   1998.0              1065.0         7290.0   
3       483.0   1805.0              1525.0         7432.0   
4       482.0   1667.0              3521.0         7494.0   

   Chronic respiratory diseases  Cirrhosis and other chronic liver diseases  \
0                        7222.0                                      3346.0   
1                        7143.0                                      3316.0   
2                        7045.0                                      3291.0   
3                        6916.0                                      3318.0   
4                        6878.0                                      3353.0   

   Digestive diseases  Fire, heat, and hot substances  Acute hepatitis  
0              6458.0                           481.0           3437.0  
1              6408.0                           462.0           3005.0  
2              6358.0                           448.0           2663.0  
3              6370.0                           448.0           2365.0  
4              6398.0                           445.0           2264.0  

[5 rows x 36 columns]

Create a column for 'Causes' and another for 'Case Count' using `.melt()`

In [43]:
id_vars = ['Entity', 'Code', 'Year']
value_vars = causesofDeath_df_copy.columns[3:]

causesofDeath_df_copy = pd.melt(causesofDeath_df_copy, id_vars=id_vars, value_vars=value_vars, 
                                var_name='Causes', value_name='Case_Count')

In [44]:
causesofDeath_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240009 entries, 0 to 240008
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Entity      240009 non-null  object
 1   Code        240009 non-null  object
 2   Year        240009 non-null  int64 
 3   Causes      240009 non-null  object
 4   Case_Count  240009 non-null  object
dtypes: int64(1), object(4)
memory usage: 9.2+ MB


In [45]:
# let's find out the values that made Case_Count an object type
objectType = []

for i in causesofDeath_df_copy.Case_Count:
      try:
            int(i)
      except:
            objectType.append(i)

In [46]:
set(objectType)

{'>1', '>1000'}

In [47]:
objectType = ['>1', '>1000']

causesofDeath_df_copy.query('Case_Count in @objectType')

Entity Code  Year                                        Causes  \
1243        China  CHN  2008  Number of executions (Amnesty International)   
1244        China  CHN  2009  Number of executions (Amnesty International)   
1245        China  CHN  2010  Number of executions (Amnesty International)   
1246        China  CHN  2011  Number of executions (Amnesty International)   
1247        China  CHN  2012  Number of executions (Amnesty International)   
1248        China  CHN  2013  Number of executions (Amnesty International)   
1249        China  CHN  2014  Number of executions (Amnesty International)   
1250        China  CHN  2015  Number of executions (Amnesty International)   
1251        China  CHN  2016  Number of executions (Amnesty International)   
3628     Malaysia  MYS  2015  Number of executions (Amnesty International)   
4494  North Korea  PRK  2013  Number of executions (Amnesty International)   
4495  North Korea  PRK  2014  Number of executions (Amnesty International)   
4496  North Korea  PRK  2015  Number of executions (Amnesty International)   
4497  North Korea  PRK  2016  Number of executions (Amnesty International)   
5842  South Sudan  SSD  2016  Number of executions (Amnesty International)   
6889      Vietnam  VNM  2015  Number of executions (Amnesty International)   
6890      Vietnam  VNM  2016  Number of executions (Amnesty International)   

     Case_Count  
1243      >1000  
1244      >1000  
1245      >1000  
1246      >1000  
1247      >1000  
1248      >1000  
1249      >1000  
1250      >1000  
1251      >1000  
3628         >1  
4494         >1  
4495         >1  
4496         >1  
4497         >1  
5842         >1  
6889         >1  
6890         >1

Let's focus on only health-related deaths

In [48]:
causesofDeath_df_copy.Causes.unique()

array(['Number of executions (Amnesty International)', 'Meningitis',
       "Alzheimer's disease and other dementias", "Parkinson's disease",
       'Nutritional deficiencies', 'Malaria', 'Drowning',
       'Interpersonal violence', 'Maternal disorders', 'HIV/AIDS',
       'Drug use disorders', 'Tuberculosis', 'Cardiovascular diseases',
       'Lower respiratory infections', 'Neonatal disorders',
       'Alcohol use disorders', 'Self', 'Exposure to forces of nature',
       'Diarrheal diseases', 'Environmental heat and cold exposure',
       'Neoplasms', 'Conflict and terrorism', 'Diabetes mellitus',
       'Chronic kidney disease', 'Poisonings', 'Protein',
       'Terrorism (deaths)', 'Road injuries',
       'Chronic respiratory diseases',
       'Cirrhosis and other chronic liver diseases', 'Digestive diseases',
       'Fire, heat, and hot substances', 'Acute hepatitis'], dtype=object)

We'll drop deaths by;
* Executions ['Number of executions (Amnesty International)']
* Drowning
* Interpersonal violence
* Alcohol use disorders
* Self
* Exposure to forces of nature
* Environmental heat and cold exposure
* Conflict and terrorism
* Poisonings
* Terrorism (deaths)
* Road injuries
* Fire, heat and hot substances

In [49]:
dropCauses = ['Number of executions (Amnesty International)', 'Drowning', 'Interpersonal violence','Alcohol use disorders', 
              'Self', 'Exposure to forces of nature', 'Environmental heat and cold exposure', 'Conflict and terrorism', 'Poisonings',
              'Terrorism (deaths)', 'Road injuries', 'Fire, heat, and hot substances']

In [50]:
dropIndx = causesofDeath_df_copy.query('Causes in @dropCauses').index

len(dropIndx)/len(causesofDeath_df_copy.Causes)

0.36363636363636365

The causes we want to drop make up 36.4% of the dataset which is okay

In [51]:
causesofDeath_df_copy = causesofDeath_df_copy.drop(index=dropIndx)

In [52]:
causesofDeath_df_copy.query('Causes in @dropCauses')

Empty DataFrame
Columns: [Entity, Code, Year, Causes, Case_Count]
Index: []

In [53]:
causesofDeath_df_copy.shape     

(152733, 5)

Convert 'Case_Count' to integer

In [54]:
causesofDeath_df_copy['Case_Count'] = causesofDeath_df_copy['Case_Count'].astype('int')

Convert 'Year' column to datetime 

In [55]:
causesofDeath_df_copy['Year'] = pd.to_datetime(causesofDeath_df_copy.Year, format='%Y')

In [56]:
causesofDeath_df_copy.isna().sum()

Entity        0
Code          0
Year          0
Causes        0
Case_Count    0
dtype: int64

In [57]:
causesofDeath_df_copy.sample(10)

Entity Code       Year                        Causes  \
96769           Finland  FIN 2005-01-01  Lower respiratory infections   
73117        Azerbaijan  AZE 1990-01-01            Drug use disorders   
132705   Eastern Europe    0 1994-01-01            Diarrheal diseases   
167934            Benin  BEN 2018-01-01        Chronic kidney disease   
174220            Wales    0 2017-01-01        Chronic kidney disease   
100358      South Korea  KOR 1991-01-01  Lower respiratory infections   
172939  Solomon Islands  SLB 2016-01-01        Chronic kidney disease   
36322            Zambia  ZMB 2007-01-01      Nutritional deficiencies   
135413      North Korea  PRK 1991-01-01            Diarrheal diseases   
168516            Chile  CHL 2015-01-01        Chronic kidney disease   

        Case_Count  
96769         1263  
73117           14  
132705           0  
167934        1543  
174220         413  
100358        4365  
172939         117  
36322         3141  
135413         971  
168516        4473

#### `ageGroupDeaths_df`
* ~~Create a subset of only African countries~~
* Create a column for 'Age Range' and another for 'Death Count' using `.melt()`

In [58]:
ageGroupDeaths_df.shape

(6840, 8)

In [59]:
ageGroupDeaths_df_copy.head()

Entity Code  Year  Deaths 70+ years  Deaths 50-69 years  \
0  Afghanistan  AFG  1990             34681               40624   
1  Afghanistan  AFG  1991             35271               41414   
2  Afghanistan  AFG  1992             35694               41829   
3  Afghanistan  AFG  1993             36433               42598   
4  Afghanistan  AFG  1994             37460               43923   

   Deaths Age: 15-49 years  Deaths 5-14 years  Deaths Under 5  
0                    22968               5642           78170  
1                    26523               6433           83035  
2                    30000               6854           93859  
3                    32339               7206          106537  
4                    36817               8106          115334

Creating subset of only African countries

In [60]:
#ageGroupDeaths_df_copy = ageGroupDeaths_df_copy.query('Code in @african_three_letter_code')

In [61]:
ageGroupDeaths_df_copy.shape

(6840, 8)

In [62]:
ageGroupDeaths_df_copy.Entity.sample(10)

2233              Germany
4148              Nigeria
1586             Dominica
6004              Tokelau
3621            Mauritius
552               Belgium
1804    Equatorial Guinea
5423              Somalia
1468              Czechia
2117                  G20
Name: Entity, dtype: object

Create a column for 'Age Range' and another for 'Death Count' using `.melt()`

In [63]:
id_vars = ['Entity', 'Code', 'Year']
value_vars = ageGroupDeaths_df_copy.columns[3:]

ageGroupDeaths_df_copy = pd.melt(ageGroupDeaths_df_copy, id_vars=id_vars, value_vars=value_vars, 
                                 var_name='Age_Group', value_name='Death_Count')

In [64]:
ageGroupDeaths_df_copy.shape

(34200, 5)

In [65]:
ageGroupDeaths_df_copy['Age_Group'].unique()

array(['Deaths 70+ years', 'Deaths 50-69 years',
       'Deaths Age: 15-49 years', 'Deaths 5-14 years', 'Deaths Under 5'],
      dtype=object)

Let's remove the 'Deaths' from the values in 'Age_Group'

In [66]:
ageGroupDeaths_df_copy['Age_Group'] = ageGroupDeaths_df_copy['Age_Group'].str.replace('Deaths ', '')
ageGroupDeaths_df_copy['Age_Group'] = ageGroupDeaths_df_copy['Age_Group'].str.replace('Age: ', '')

In [67]:
ageGroupDeaths_df_copy['Age_Group'].unique()

array(['70+ years', '50-69 years', '15-49 years', '5-14 years', 'Under 5'],
      dtype=object)

Convert 'Year' to datetime

In [68]:
ageGroupDeaths_df_copy['Year'] = pd.to_datetime(ageGroupDeaths_df_copy.Year, format='%Y')

In [69]:
ageGroupDeaths_df_copy.sample(10)

Entity Code       Year    Age_Group  Death_Count
14846             China  CHN 2016-01-01  15-49 years       936626
32652         Singapore  SGP 2002-01-01      Under 5          156
4799              Qatar  QAT 2019-01-01    70+ years          939
28734           Croatia  HRV 2014-01-01      Under 5          186
20905        Azerbaijan  AZE 2015-01-01   5-14 years          623
18981         Singapore  SGP 2011-01-01  15-49 years         1717
24833  Northern Ireland  NaN 2013-01-01   5-14 years           25
32894       South Sudan  SSD 2004-01-01      Under 5        46436
20014     United States  USA 1994-01-01  15-49 years       255340
9933               Laos  LAO 1993-01-01  50-69 years        11355

#### `doctorsPerPopulation_df`
* ~~Create subset of only African countries~~

In [70]:
doctorsPerPopulation_df_copy.sample(5)

IndicatorCode                                     Indicator  \
10816      HWF_0005  Medical doctors not further defined (number)   
576        HWF_0001                  Medical doctors (per 10,000)   
3189       HWF_0002                      Medical doctors (number)   
5096       HWF_0002                      Medical doctors (number)   
9640       HWF_0005  Medical doctors not further defined (number)   

      ParentLocationCode ParentLocation ThreeLocCode            Location  \
10816                EUR         Europe          RUS  Russian Federation   
576                  AMR       Americas          ECU             Ecuador   
3189                 AFR         Africa          TCD                Chad   
5096                 EUR         Europe          LVA              Latvia   
9640                 EUR         Europe          AUT             Austria   

       Period  FactValueNumeric    Value  
10816    2006         702200.00  702 200  
576      1997             13.28    13.28  
3189     2019            859.00      859  
5096     2017           6225.00     6225  
9640     2016           7168.00     7168

In [71]:
doctorsPerPopulation_df_copy.shape

(11508, 9)

In [72]:
#doctorsPerPopulation_df_copy = doctorsPerPopulation_df_copy.query('ThreeLocCode in @african_three_letter_code')

In [73]:
doctorsPerPopulation_df_copy.shape

(11508, 9)

Let's remove the whitespace characters in 'Value'

In [74]:
doctorsPerPopulation_df_copy['Value'] = doctorsPerPopulation_df_copy.Value.str.replace('\xa0', '')

In [75]:
objectType = []

for i in doctorsPerPopulation_df_copy.Value:
      try:
            float(i)
      except:
            objectType.append(i)

In [76]:
set(objectType)

set()

Change the column 'Period' to 'Year' then convert to datetime.

In [77]:
doctorsPerPopulation_df_copy.rename(columns={'Period':'Year'}, inplace=True)

doctorsPerPopulation_df_copy['Year'] = pd.to_datetime(doctorsPerPopulation_df_copy.Year, format='%Y')

In [78]:
doctorsPerPopulation_df_copy.sample(15)

IndicatorCode                                     Indicator  \
7484      HWF_0003     Generalist medical practitioners (number)   
8964      HWF_0005  Medical doctors not further defined (number)   
4979      HWF_0002                      Medical doctors (number)   
666       HWF_0001                  Medical doctors (per 10,000)   
4178      HWF_0002                      Medical doctors (number)   
4684      HWF_0002                      Medical doctors (number)   
3929      HWF_0002                      Medical doctors (number)   
6868      HWF_0003     Generalist medical practitioners (number)   
3267      HWF_0002                      Medical doctors (number)   
4952      HWF_0002                      Medical doctors (number)   
7106      HWF_0003     Generalist medical practitioners (number)   
5558      HWF_0002                      Medical doctors (number)   
2879      HWF_0002                      Medical doctors (number)   
3815      HWF_0002                      Medical doctors (number)   
3354      HWF_0002                      Medical doctors (number)   

     ParentLocationCode         ParentLocation ThreeLocCode  \
7484                WPR        Western Pacific          AUS   
8964                AFR                 Africa          NGA   
4979                EUR                 Europe          KGZ   
666                 AMR               Americas          MEX   
4178                EUR                 Europe          CHE   
4684                EUR                 Europe          HRV   
3929                EUR                 Europe          AND   
6868                EUR                 Europe          GRC   
3267                AMR               Americas          ATG   
4952                EUR                 Europe          KAZ   
7106                EUR                 Europe          MDA   
5558                EUR                 Europe          SWE   
2879                AFR                 Africa          BWA   
3815                EMR  Eastern Mediterranean          PAK   
3354                AMR               Americas          COL   

                 Location       Year  FactValueNumeric  Value  
7484            Australia 1980-01-01          18146.00  18146  
8964              Nigeria 2009-01-01          58363.00  58363  
4979           Kyrgyzstan 2004-01-01          12905.00  12905  
666                Mexico 1994-01-01             15.65  15.65  
4178          Switzerland 2002-01-01          25921.00  25921  
4684              Croatia 2002-01-01          10589.00  10589  
3929              Andorra 2001-01-01            175.00    175  
6868               Greece 2011-01-01           3369.00   3369  
3267  Antigua and Barbuda 2017-01-01            264.00    264  
4952           Kazakhstan 1996-01-01          57941.00  57941  
7106  Republic of Moldova 2003-01-01           2446.00   2446  
5558               Sweden 2005-01-01          31199.00  31199  
2879             Botswana 2018-01-01            853.00    853  
3815             Pakistan 1998-01-01          82682.00  82682  
3354             Colombia 1992-01-01          31113.00  31113

Since the 'FactValueNumeric' and the 'Value' column are the same, let's drop the 'Value' column.

In [79]:
doctorsPerPopulation_df_copy = doctorsPerPopulation_df_copy.drop('Value', axis=1)

#### `healthExpenditure`
* ~~Create subset of African countries~~
* Drop 'Indicator Name'
* Create a column for 'Year' and 'Health Expenditure (% of GDP)' for values using `.melt()`

In [80]:
healthExpenditure_copy.sample(5)

Country Name Country Code                         Indicator Name  \
145               Latvia          LVA  Current health expenditure (% of GDP)   
72              Ethiopia          ETH  Current health expenditure (% of GDP)   
206                Sudan          SDN  Current health expenditure (% of GDP)   
125  St. Kitts and Nevis          KNA  Current health expenditure (% of GDP)   
243              Tunisia          TUN  Current health expenditure (% of GDP)   

         2000      2001      2002      2003      2004      2005      2006  \
145  5.439958  5.777623  5.770019  5.489841  6.216779  5.847668  5.718347   
72   4.365047  4.723167  4.712630  4.899709  4.312819  4.100981  4.457573   
206  3.606645  3.364568  3.300109  3.891879  4.155935  4.094181  4.885443   
125  4.799338  4.396937  4.452106  4.751597  4.490674  4.728039  4.982034   
243  5.048344  5.099295  5.370927  5.463785  5.453238  5.397361  5.382182   

     ...      2012      2013      2014      2015      2016      2017  \
145  ...  5.380903  5.367528  5.466149  5.654695  6.135920  5.970317   
72   ...  4.539596  4.075065  4.033636  3.823171  3.600331  3.453362   
206  ...  5.688270  6.964875  5.676197  7.267471  5.484100  6.055461   
125  ...  5.172269  5.260129  4.940392  5.313013  5.632156  5.596308   
243  ...  6.606423  6.856620  6.807396  6.621541  6.609990  6.849278   

         2018      2019  2020  2021  
145  6.191036  6.578142   NaN   NaN  
72   3.308187  3.238086   NaN   NaN  
206  4.436535  4.573566   NaN   NaN  
125  5.212878  5.396328   NaN   NaN  
243  6.820991  6.958459   NaN   NaN  

[5 rows x 25 columns]

In [81]:
healthExpenditure_copy.shape

(266, 25)

Create subset of African countries

In [82]:
# change Country Code column name
healthExpenditure_copy.rename(columns={'Country Code':'Country_Code'}, inplace=True)
#healthExpenditure_copy = healthExpenditure_copy.query('Country_Code in @african_three_letter_code')

Create a column for 'Year' and 'Health Expenditure (% of GDP)' for values using `.melt()` and drop 'Indicator Name'

In [83]:
id_vars = ['Country Name', 'Country_Code']
value_vars = healthExpenditure_copy.columns[3:]

healthExpenditure_copy = pd.melt(healthExpenditure_copy, id_vars=id_vars, value_vars=value_vars, 
                                 var_name='Year', value_name='Health Expenditure (% of GDP)')

Convert 'Year' column to datetime 

In [84]:
healthExpenditure_copy['Year'] = pd.to_datetime(healthExpenditure_copy.Year, format='%Y')

In [85]:
healthExpenditure_copy.shape

(5852, 4)

In [86]:
healthExpenditure_copy.sample(10)

Country Name Country_Code       Year  \
5547   Syrian Arab Republic          SYR 2020-01-01   
2075                Somalia          SOM 2007-01-01   
5133  Micronesia, Fed. Sts.          FSM 2019-01-01   
2865           Saudi Arabia          SAU 2010-01-01   
4818               Barbados          BRB 2018-01-01   
5634             Costa Rica          CRI 2021-01-01   
443                  Norway          NOR 2001-01-01   
2789                Lao PDR          LAO 2010-01-01   
3168    Trinidad and Tobago          TTO 2011-01-01   
1010             San Marino          SMR 2003-01-01   

      Health Expenditure (% of GDP)  
5547                            NaN  
2075                            NaN  
5133                      11.442080  
2865                       3.647856  
4818                       6.562592  
5634                            NaN  
443                        8.022024  
2789                       2.911552  
3168                       4.721979  
1010                       4.815475

### Storing data

In [87]:
'''
countryCodes_df_copy.to_excel('./data/clean_countryCodes.xlsx', index=False)
causesofDeath_df_copy.to_excel('./data/clean_causesofDeath.xlsx', index=False)
ageGroupDeaths_df_copy.to_excel('./data/clean_ageGroupDeaths.xlsx', index=False)
doctorsPerPopulation_df_copy.to_excel('./data/clean_doctorsPerPopulation.xlsx', index=False)
healthExpenditure_copy.to_excel('./data/clean_healthExpenditure.xlsx', index=False)
africanCountries.to_excel('./data/clean_africanCountries.xlsx', index=False)
'''

"\ncountryCodes_df_copy.to_excel('./data/clean_countryCodes.xlsx', index=False)\ncausesofDeath_df_copy.to_excel('./data/clean_causesofDeath.xlsx', index=False)\nageGroupDeaths_df_copy.to_excel('./data/clean_ageGroupDeaths.xlsx', index=False)\ndoctorsPerPopulation_df_copy.to_excel('./data/clean_doctorsPerPopulation.xlsx', index=False)\nhealthExpenditure_copy.to_excel('./data/clean_healthExpenditure.xlsx', index=False)\nafricanCountries.to_excel('./data/clean_africanCountries.xlsx', index=False)\n"

In [88]:
# We then group the dataframes into a list for more efficient processing 
dflist = [countryCodes_df_copy, causesofDeath_df_copy, ageGroupDeaths_df_copy,
          doctorsPerPopulation_df_copy, healthExpenditure_copy, africanCountries]
sheetname = ['World_Countries', 'Causes_of_Death', 'Death_by_Age_Group',
             'Doctors_per_Population', 'Health_Expenditure', 'African_Countries']

# We'll define an Excel writer object and the target file
Excelwriter = pd.ExcelWriter("./data/World Healthcaredata.xlsx",engine="xlsxwriter")

#We now we'll loop the list of dataframes
for i, df in enumerate (dflist):
    df.to_excel(Excelwriter, sheet_name=sheetname[i],index=False)
#And finally we save the file
Excelwriter.save()